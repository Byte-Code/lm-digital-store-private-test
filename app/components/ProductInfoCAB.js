import React, { Component, PropTypes } from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
import styled from 'styled-components';
import { List } from 'immutable';
import AddIcon from 'material-ui/svg-icons/content/add-circle-outline';
import RemoveIcon from 'material-ui/svg-icons/content/remove-circle-outline';

import { titleFormatter } from '../utils/utils';
import Card from './Card';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

const Row = styled.div`
  width: 100%;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  background-color: ${props => props.bgColor}
  padding-top: ${props => props.pTop};
  padding-left: 125px;
  padding-right: 125px;
  padding-bottom: 57px;
`;

const Column = styled.div`
  width: calc((100% - 100px) /2);
`;

const Section = styled.div`
  margin-bottom: 20px;
  &>p{
    font-size: ${props => props.fSize};
    font-family: LeroyMerlinSans Light;
    &>span{
      font-family: LeroyMerlinSans;
    }
  }
`;

const CardTitle = styled.div`
  display: flex;
  cursor: pointer;
  align-items: center;
  padding: 19px 125px;
  background: #f7f7f7;
  &>h3{
    margin-left: 20px;
    font-size: 16px;
    line-height: 32px;
  }
`;

export const Title = styled.div`
  font-family: LeroyMerlinSans Italic;
  font-size: ${props => props.fSize};
  line-height: 20px;
  color: #339900;
  margin-bottom: 14px;
`;

const Divider = styled.div`
  border: 1px dashed #67cb33;
  width: 100%;
  margin-top: 20px;
`;

function CardTitleComponent({ expanded }) {
  const Icon = expanded ? RemoveIcon : AddIcon;
  return (
    <CardTitle>
      <Icon color="#333333" />
      <h3>Maggiori informazioni</h3>
    </CardTitle>
  );
}

CardTitleComponent.propTypes = {
  expanded: PropTypes.bool.isRequired
};

export default class ProductInfoCAB extends Component {
  static propTypes = {
    marketingDescriptions: ImmutablePropTypes.map.isRequired,
    descriptions: ImmutablePropTypes.seq.isRequired
  }

  renderBlocks() {
    const { marketingDescriptions } = this.props;
    if (!marketingDescriptions) {
      return null;
    }
    // TODO why can this be undefined?
    const blocks = marketingDescriptions.get('chooseBlocks') || List();

    return blocks.map(block => (
      <Column key={block.get('title')}>
        <Section fSize="18px">
          <Title fSize="20px">{titleFormatter(block.get('title'))}</Title>
          {block.get('customerChooses').map(c =>
            <p key={c.get('description')}>{c.get('description')}</p>)}
        </Section>
        <Divider />
      </Column>
    ));
  }

  renderDescriptions(index) {
    const { descriptions } = this.props;
    const currentBatch = descriptions.get(index);
    if (!currentBatch) {
      return null;
    }

    return currentBatch.map(desc => (
      <Section fSize="14px" key={desc.get('code')}>
        <Title fSize="16px">{titleFormatter(desc.get('title'))}</Title>
        {desc.get('description').map(item => (
          <p key={`${item.get('label')}${item.get('value')}`}>
            <span>{item.get('label') && `${item.get('label')}: `}</span>
            {item.get('value')}
          </p>
        ))}
        {(descriptions.get(index).last() !== desc) && <Divider />}
      </Section>
    ));
  }

  render() {
    return (
      <Wrapper>
        <Row pTop="69px" bgColor="#efefef">
          {this.renderBlocks()}
        </Row>
        <Card TitleComponent={CardTitleComponent}>
          <Row pTop="31px" bgColor="#f7f7f7">
            <Column>
              {this.renderDescriptions(0)}
            </Column>
            <Column>
              {this.renderDescriptions(1)}
            </Column>
          </Row>
        </Card>
      </Wrapper>
    );
  }
}
