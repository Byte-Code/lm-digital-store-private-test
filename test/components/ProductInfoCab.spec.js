import React from 'react';
import { shallow } from 'enzyme';
import { Map, Seq, fromJS } from 'immutable';

import ProductInfoCAB, { Title } from '../../app/components/ProductInfoCAB';

const emptyDescriptions = Seq();
const emptyMarketingDescriptions = Map();
const marketingDescriptions = fromJS({
  chooseBlocks: [
    { title: 'desc0',
      customerChooses: [{ description: 'foo' }, { description: 'bar' }]
    },
    { title: 'desc1',
      customerChooses: [{ description: 'foobar' }]
    }
  ]
});
const descriptions = fromJS([
  [{
    code: 'description0',
    title: 'description0',
    description: [
      { label: 'label00', value: 'value00' },
      { label: 'label01', value: 'value01' },
      { label: 'label02', value: 'value02' }
    ]
  },
  {
    code: 'description01',
    title: 'description01',
    description: [
      { label: 'label03', value: 'value03' },
    ]
  },
  ],
  [{
    code: 'description1',
    title: 'description1',
    description: [
      { label: 'label04', value: 'value04' },
      { label: 'label05', value: 'value05' },
      { label: 'label06', value: 'value06' }
    ]
  }],
  [{
    code: 'descriptionNotRendered',
    title: 'descriptionNotRendered',
    description: [
      { label: 'labelNR', value: 'valueNR' }
    ]
  }]
]).toSeq();

describe('ProductInfoCAB', () => {
  it('should render no descriptions when they are not provided', () => {
    const result = shallow(
      <ProductInfoCAB
        descriptions={emptyDescriptions}
        marketingDescriptions={emptyMarketingDescriptions}
      />
    );
    expect(result).toMatchSnapshot();
  });

  it('should render a list of marketingDescriptions', () => {
    const result = shallow(
      <ProductInfoCAB
        descriptions={emptyDescriptions}
        marketingDescriptions={marketingDescriptions}
      />
    );
    expect(result.find(Title)).toHaveLength(2);
    expect(result.find('p')).toHaveLength(3);
    expect(result).toMatchSnapshot();
  });

  it('should render a max of two list of descriptions', () => {
    const result = shallow(
      <ProductInfoCAB
        descriptions={descriptions}
        marketingDescriptions={emptyMarketingDescriptions}
      />
    );
    expect(result.find(Title)).toHaveLength(3);
    expect(result.find('p')).toHaveLength(7);
    expect(result).toMatchSnapshot();
  });

  it('should render both descriptions and marketingDescriptions', () => {
    const result = shallow(
      <ProductInfoCAB
        descriptions={descriptions}
        marketingDescriptions={marketingDescriptions}
      />
    );
    expect(result.find(Title)).toHaveLength(5);
    expect(result.find('p')).toHaveLength(10);
    expect(result).toMatchSnapshot();
  });
});
