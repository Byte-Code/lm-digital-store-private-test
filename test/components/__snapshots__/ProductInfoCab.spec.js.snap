// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductInfoCAB should render a list of marketingDescriptions 1`] = `
<glamorous(div)>
  <glamorous(div)
    backgroundColor="#efefef"
    paddingTop="69px"
  >
    <glamorous(div)>
      <glamorous(div)
        fontSize="18px"
        lineHeight="28px"
      >
        <glamorous(div)
          fontSize="20px"
        >
          desc0
        </glamorous(div)>
        <p>
          foo
        </p>
        <p>
          bar
        </p>
      </glamorous(div)>
      <glamorous(div) />
    </glamorous(div)>
    <glamorous(div)>
      <glamorous(div)
        fontSize="18px"
        lineHeight="28px"
      >
        <glamorous(div)
          fontSize="20px"
        >
          desc1
        </glamorous(div)>
        <p>
          foobar
        </p>
      </glamorous(div)>
      <glamorous(div) />
    </glamorous(div)>
  </glamorous(div)>
  <Card
    TitleComponent={[Function]}
  >
    <glamorous(div)
      Card="31px"
      backgroundColor="#f7f7f7"
    >
      <glamorous(div) />
      <glamorous(div) />
    </glamorous(div)>
  </Card>
</glamorous(div)>
`;

exports[`ProductInfoCAB should render a max of two list of descriptions 1`] = `
<glamorous(div)>
  <glamorous(div)
    backgroundColor="#efefef"
    paddingTop="69px"
  />
  <Card
    TitleComponent={[Function]}
  >
    <glamorous(div)
      Card="31px"
      backgroundColor="#f7f7f7"
    >
      <glamorous(div)>
        <glamorous(div)
          fontSize="14px"
          lineHeight="20px"
        >
          <glamorous(div)
            fontSize="16px"
          >
            description0
          </glamorous(div)>
          <p>
            <span>
              label00: 
            </span>
            value00
          </p>
          <p>
            <span>
              label01: 
            </span>
            value01
          </p>
          <p>
            <span>
              label02: 
            </span>
            value02
          </p>
          <glamorous(div) />
        </glamorous(div)>
        <glamorous(div)
          fontSize="14px"
          lineHeight="20px"
        >
          <glamorous(div)
            fontSize="16px"
          >
            description01
          </glamorous(div)>
          <p>
            <span>
              label03: 
            </span>
            value03
          </p>
        </glamorous(div)>
      </glamorous(div)>
      <glamorous(div)>
        <glamorous(div)
          fontSize="14px"
          lineHeight="20px"
        >
          <glamorous(div)
            fontSize="16px"
          >
            description1
          </glamorous(div)>
          <p>
            <span>
              label04: 
            </span>
            value04
          </p>
          <p>
            <span>
              label05: 
            </span>
            value05
          </p>
          <p>
            <span>
              label06: 
            </span>
            value06
          </p>
        </glamorous(div)>
      </glamorous(div)>
    </glamorous(div)>
  </Card>
</glamorous(div)>
`;

exports[`ProductInfoCAB should render both descriptions and marketingDescriptions 1`] = `
<glamorous(div)>
  <glamorous(div)
    backgroundColor="#efefef"
    paddingTop="69px"
  >
    <glamorous(div)>
      <glamorous(div)
        fontSize="18px"
        lineHeight="28px"
      >
        <glamorous(div)
          fontSize="20px"
        >
          desc0
        </glamorous(div)>
        <p>
          foo
        </p>
        <p>
          bar
        </p>
      </glamorous(div)>
      <glamorous(div) />
    </glamorous(div)>
    <glamorous(div)>
      <glamorous(div)
        fontSize="18px"
        lineHeight="28px"
      >
        <glamorous(div)
          fontSize="20px"
        >
          desc1
        </glamorous(div)>
        <p>
          foobar
        </p>
      </glamorous(div)>
      <glamorous(div) />
    </glamorous(div)>
  </glamorous(div)>
  <Card
    TitleComponent={[Function]}
  >
    <glamorous(div)
      Card="31px"
      backgroundColor="#f7f7f7"
    >
      <glamorous(div)>
        <glamorous(div)
          fontSize="14px"
          lineHeight="20px"
        >
          <glamorous(div)
            fontSize="16px"
          >
            description0
          </glamorous(div)>
          <p>
            <span>
              label00: 
            </span>
            value00
          </p>
          <p>
            <span>
              label01: 
            </span>
            value01
          </p>
          <p>
            <span>
              label02: 
            </span>
            value02
          </p>
          <glamorous(div) />
        </glamorous(div)>
        <glamorous(div)
          fontSize="14px"
          lineHeight="20px"
        >
          <glamorous(div)
            fontSize="16px"
          >
            description01
          </glamorous(div)>
          <p>
            <span>
              label03: 
            </span>
            value03
          </p>
        </glamorous(div)>
      </glamorous(div)>
      <glamorous(div)>
        <glamorous(div)
          fontSize="14px"
          lineHeight="20px"
        >
          <glamorous(div)
            fontSize="16px"
          >
            description1
          </glamorous(div)>
          <p>
            <span>
              label04: 
            </span>
            value04
          </p>
          <p>
            <span>
              label05: 
            </span>
            value05
          </p>
          <p>
            <span>
              label06: 
            </span>
            value06
          </p>
        </glamorous(div)>
      </glamorous(div)>
    </glamorous(div)>
  </Card>
</glamorous(div)>
`;

exports[`ProductInfoCAB should render no descriptions when they are not provided 1`] = `
<glamorous(div)>
  <glamorous(div)
    backgroundColor="#efefef"
    paddingTop="69px"
  />
  <Card
    TitleComponent={[Function]}
  >
    <glamorous(div)
      Card="31px"
      backgroundColor="#f7f7f7"
    >
      <glamorous(div) />
      <glamorous(div) />
    </glamorous(div)>
  </Card>
</glamorous(div)>
`;
